core/commands/pin.go

```diff
@@ -594,7 +596,10 @@ type pinVerifyOpts struct {
 
 func pinVerify(ctx context.Context, n *core.IpfsNode, opts pinVerifyOpts) <-chan interface{} {
 	visited := make(map[string]PinStatus)
-	getLinks := n.DAG.GetOfflineLinkService().GetLinks
+
+	bs := n.Blocks.Blockstore()
+	DAG := dag.NewDAGService(bserv.New(bs, offline.Exchange(bs)))
+	getLinks := dag.GetLinksWithDAG(DAG)
```

Comment:
This is the new way to do offline stuff. We could also use the context trick if we get tired of this dance.

Reply:
this is 3 lines instead of 1; ergonomically it's not an improvement...

Reply:
Discussion: https://github.com/ipfs/go-ipfs/issues/4009

So, first, I made implementing `LinkService` optional. If your `NodeGetter` (the get side of the `DAGService`) doesn't implement the `LinkService` optimizations, it doesn't have to implement the interface.

Next, I tried adding a `OfflineNodeGetter` method to the `NodeGetter` interface. However, that really tricky and messy:

1. We realized that we'd probably want `Offline*` methods on every related service (which would kind of suck).
2. Things like bitswap sessions (which I wanted to turn into valid `NodeGetter`s would *also* have to implement this method). I simply couldn't find a clean way to do this.

So, in #4009, I suggested adding an "offline" flag to the context. I *still* think that's the best way to go because I keep running into similar cases where I want to try to do something without touching the network but @kevina reasonably objected that it was a bit magical and easy to screw up.

Hence this solution.

I could add a helper function but, really, I kind of want this to be ugly. I'm not happy with this solution at all and don't want this pattern repeated all over the codebase. For one, it makes the `LinkService` optimization totally useless because we don't *reuse* the DAG (we construct a new one with the blockstore when we do the GC).

Reply:
fair enough.